pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,338 @@
+
+	return nil
+}
+
+func calculateDaemonSetOverhead(pods []v1.Pod) v1.ResourceList {
+	overheads := v1.ResourceList{}
+	// calculating daemonset overhead
+	daemonSetPods := []v1.Pod{}
+	for _, pod := range pods {
+		if pod.GetOwnerReferences()[0].Kind == "DaemonSet" {
+			daemonSetPods = append(daemonSetPods, pod)
+		}
+	}
+	daemonSetRequests, _ := getPodsTotalRequestsAndLimits(daemonSetPods)
+	addResourceQuantity(daemonSetRequests, overheads)
+	return overheads
+}
+
+func (c *Controller) insertGaugeValues(resources map[v1.ResourceName]resource.Quantity, prefix string, labels *prometheus.Labels) error {
+	for resourceName, quantity := range resources {
+		gaugeVecKey := prefix + "_" + strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+		gaugeVec, ok := c.GaugeVecMap[gaugeVecKey]
+		if !ok {
+			gaugeVec = prometheus.NewGaugeVec(
+				prometheus.GaugeOpts{
+					Namespace: "karpenter",
+					Subsystem: "nodes",
+					Name:      gaugeVecKey,
+					Help:      "Auto-generated gauge: " + gaugeVecKey,
+				},
+				getLabelNames(),
+			)
+			c.GaugeVecMap[gaugeVecKey] = gaugeVec
+			crmetrics.Registry.MustRegister(gaugeVec)
+		}
+		gauge, err := gaugeVec.GetMetricWith(*labels)
+		if err != nil {
+			return fmt.Errorf("generate new gauge: %w", err)
+		}
+		if resourceName == v1.ResourceCPU {
+			gauge.Set(float64(quantity.MilliValue()))
+		} else {
+			gauge.Set(float64(quantity.Value()))
+		}
+	}
+	return nil
+}
+
+// GetPodsTotalRequestsAndLimits calculates the total resource requests and limits for the pods.
+// If pod overhead is non-nil, the pod overhead is added to the
+// total container resource requests and to the total container limits which have a non-zero quantity.
+func getPodsTotalRequestsAndLimits(pods []v1.Pod) (reqs map[v1.ResourceName]resource.Quantity, limits map[v1.ResourceName]resource.Quantity) {
+	reqs, limits = map[v1.ResourceName]resource.Quantity{}, map[v1.ResourceName]resource.Quantity{}
+	for _, pod := range pods {
+		// Excluding pods that are completed or failed
+		if pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodSucceeded {
+			continue
+		}
+		for _, container := range pod.Spec.Containers {
+			// Calculate Resource Requests
+			addResourceQuantity(container.Resources.Requests, reqs)
+			// Calculate Resource Limits
+			addResourceQuantity(container.Resources.Limits, limits)
+		}
+		// Add overhead for running a pod to the sum of requests and to non-zero limits:
+		if pod.Spec.Overhead != nil {
+			// Calculate Resource Requests
+			addResourceQuantity(pod.Spec.Overhead, reqs)
+			// Calculate Resource Requests
+			// Add to limits only when non-zero
+			for resourceName, quantity := range pod.Spec.Overhead {
+				if value, ok := limits[resourceName]; ok && !value.IsZero() {
+					value.Add(quantity)
+					limits[resourceName] = value
+				}
+			}
+		}
+	}
+	return
+}
+
+func addResourceQuantity(resources v1.ResourceList, resourceMap map[v1.ResourceName]resource.Quantity) {
```

Comment:
Do we even need to record this at the node level? If we record it at the pod level, can't we just sum the metric over pods with a specific name?

Reply:
In my opinion resource level should be a node level metric instead of a pod level. We can definitely record the pod resource level but that will introduce huge loads to Prometheus for both scraping and when query hits. Aggregating data at the query time is gonna make the dashboard unusable.

The only advantage to record the pod level resource consumption is that we can break the node's workload down and identify which pods are causing problem. But I would rather the customer to figure that out using kubectl.